### ISSUE TITLE
#include <fstream> is missing in tools/solidityUpgrade/SourceUpgrade.cpp for std::ofstream

### ISSUE THREAD

**Issue by yurivict** [ethereum/solidity#13091]
std::ofstream is used there but because there's no ```#include <fstream>``` clang-13 fails on FreeBSD 13.1

**Comment by cameel** [comment]
Thanks for the report. Would you be willing to submit a PR to fix this?

I wonder if it's generally reproducible with clang or specific to FreeBSD. We do get reports about these missing includes on some platforms from time to time. Would be best if we could catch more of them in our CI.

**Comment by yurivict** [comment]
> I wonder if it's generally reproducible with clang or specific to FreeBSD. 

This issue probably depends on the c++ library in use.

**Comment by yurivict** [comment]
https://github.com/ethereum/solidity/pull/13094

**Comment by yurivict** [comment]
In general I always include all headers that are needed in each module. You seem to rely on include statements in the other headers that you include.

**Comment by cameel** [comment]
> This issue probably depends on the c++ library in use.

Yeah. I mean, I thought that clang just comes with libc++ but I checked that now and looks like I have to install it separately and also tell clang to use it. So maybe we could actually add a libc++ build to CI.

> In general I always include all headers that are needed in each module. You seem to rely on include statements in the other headers that you include.

I do that too, but not everyone does and we don't have a hard rule about that in the project so it's a bit of a mess. It's also hard to enforce automatically. We tried to add include-what-you-use as a CI check but that has too many false-positives.
